src/test/java/com/atlassian/db/replica/internal/state/ConnectionStateTest.java

```diff
@@ -163,4 +165,87 @@ public void shouldChangeStateFromMainToClosed() throws SQLException {
         verify(stateListener).transition(REPLICA, CLOSED);
     }
 
+    @Test
+    public void shouldChangeStateFromNoConnectionToReadFromMainConnection() throws SQLException {
+        final StateListener stateListener = mock(StateListener.class);
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            permanentInconsistency().build()
+        ).stateListener(stateListener)
+            .build();
+
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        verify(stateListener).transition(NOT_INITIALISED, READ_FROM_MAIN);
+    }
+
+    @Test
+    public void shouldChangeStateFromReplicaConnectionToReadFromMainConnection() throws SQLException {
+        final StateListener stateListener = mock(StateListener.class);
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            new CircularConsistency.Builder(ImmutableList.of(true, false)).build()
+        ).stateListener(stateListener)
+            .build();
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+        Mockito.reset(stateListener);
+
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        verify(stateListener).transition(REPLICA, READ_FROM_MAIN);
+    }
+
+    @Test
+    public void shouldChangeStateFromReadFromMainConnectionToReplicaConnection() throws SQLException {
+        final StateListener stateListener = mock(StateListener.class);
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
```

Comment:
Can we put those in `@Before` to unclutter the test cases?

Reply:
For me, it's easier to read code when there's no mutable class lever variable, but I can add before if you like it more. I have no strong feelings.

Reply:
Fair enough, for now let's optimize for state tracking. I have some more ideas on how to optimize for readability, so it will be a better ROI later.